src/search/pdbs/cegar.h

```diff
@@ -0,0 +1,170 @@
+#ifndef PDBS_CEGAR_H
+#define PDBS_CEGAR_H
+
+#include "pattern_collection_information.h"
+#include "pattern_database.h"
+
+#include "../task_proxy.h"
+
+#include "../utils/logging.h"
+
+#include <memory>
+#include <unordered_set>
+#include <vector>
+
+namespace options {
+class OptionParser;
+}
+
+namespace utils {
+class CountdownTimer;
+class RandomNumberGenerator;
+}
+
+namespace pdbs {
+/*
+  This is used as a "collection entry" in the CEGAR algorithm. It stores
+  the PDB (and with that, the pattern) and an optimal plan (in the wildcard
+  format) for that PDB if it exists or unsolvable is true otherwise. It can
+  be marked as "solved" to ignore it in further iterations of the CEGAR
+  algorithm.
+*/
+class Projection {
+    std::shared_ptr<PatternDatabase> pdb;
+    std::vector<std::vector<OperatorID>> plan;
+    bool unsolvable;
+    bool solved;
+
+public:
+    Projection(
+        const std::shared_ptr<PatternDatabase> &&pdb,
+        const std::vector<std::vector<OperatorID>> &&plan,
+        bool unsolvable)
+        : pdb(move(pdb)),
+          plan(move(plan)),
+          unsolvable(unsolvable),
+          solved(false) {}
+
+    const std::shared_ptr<PatternDatabase> &get_pdb() const {
+        return pdb;
+    }
+
+    const Pattern &get_pattern() const {
+        return pdb->get_pattern();
+    }
+
+    const std::vector<std::vector<OperatorID>> &get_plan() const {
+        return plan;
+    }
+
+    bool is_unsolvable() const {
+        return unsolvable;
+    }
+
+    void mark_as_solved() {
+        solved = true;
+    }
+
+    bool is_solved() {
+        return solved;
+    }
+};
+
+struct Flaw {
+    int collection_index;
+    int variable;
+
+    Flaw(int collection_index, int variable)
+        : collection_index(collection_index),
+          variable(variable) {
+    }
+};
+
+using FlawList = std::vector<Flaw>;
+
+/*
+  This class implements the CEGAR algorithm for computing pattern collections.
+
+  In a nutshell, it receives a concrete task plus a (sub)set of its goal in a
+  randomized order. Starting from the pattern collection consisting of a
+  singleton pattern for each goal variable, it repeatedly attempts to execute
+  an optimal plan of each pattern in the concrete task, collects reasons if
+  this is not possible (so-called flaws) and refines the pattern in question
+  by adding a variable to it.
+
+  Further parameters allow blacklisting a (sub)set of the non-goal variables
+  which are then never added to the collection, limiting PDB and collection
+  size, setting a time limit and switching between computing regular or
+  wildcard plans, where the latter are sequences of parallel operators
+  inducing the same abstract transition.
+*/
+class CEGAR {
+    const int max_pdb_size;
+    const int max_collection_size;
+    const bool wildcard_plans;
```

Comment:
If anything, it makes the application of the plan a bit slower because we need to test all equivalent operators in each step. But this certainly wouldn't be a bottle neck. I would prefer to keep the option, though, because we evaluate it in the paper.